### PR TITLE
Add support for noexcept keyword to embind. 

### PR DESCRIPTION
**Pull request by brendandahl**
This continues the work from #15273 and #10613 which supports the noexcept
keyword by ignoring it on c++17 compilers. I've also added a missing
template for class methods and tests for each possible use of noexcept.

### CODE REVIEW

**Review by kripken — APPROVED** [review]
I don't fully understand the C++ parts, I admit :smile: but it looks reasonable. Testing looks good.

@sbc100 maybe you also want to take a look?

**Comment by kripken** [comment]
Code size failures on CI might be fixed by merging latest `main`.

**Comment by brendandahl** [comment]
I'm not really sure why the code size would change. I'm guessing that's pretty common when changing embind.h though. Should I do the suggested `If this is expected, rerun the test with --rebaseline to update the expected sizes`?

**Comment by kripken** [comment]
Oh, I think the code size change is unrelated. We just landed a PR to fix it now on `main`, so merging that here should fix it.

Sorry for the annoyance. The issue is that we don't run code size tests on the chromium rollers, because if we did then any LLVM change that changes expectations would break the rolls, which then needs to be fixed in a multi-step process. Instead, we skip code size tests there, and then we can see the effects here on github, where the fix is pretty simple (land a change after doing `--rebaseline`). But meanwhile, all PRs will show such errors. Usually that's fairly rare, at least...

File: tests/test_other.py [hunk 1]
```diff
@@ -2499,6 +2499,7 @@ def test_embind_closure_no_dynamic_execution(self):
   def test_embind(self, extra_args):
     test_cases = [
       (['-lembind']),
+      (['-lembind', '-std=c++17']),
```
Comment (sbc100) [inline tests/test_other.py:2502]: It looks like `noexcept` was added in C++11 .. so why is this needed? https://en.cppreference.com/w/cpp/language/noexcept_spec

Comment (brendandahl) [reply tests/test_other.py:2502]: The keyword was added, but it didn't become part of the function signature until c++17.  If the changes in bind.h are removed the test will fail to compile under c++17.

Comment (sbc100) [reply tests/test_other.py:2502]: Maybe worth adding a quick comment on the line above?

**Comment by david-fong** [comment]
Thanks everyone for working on this. I'm quite happy that this got merged! Just curious about why this wasn't included in the changelog entry for 3.1.14. Was it considered too small a fish to mention?

**Comment by kripken** [comment]
@david-fong I think it's worth mentioning, good point. A PR for that would be appreciated!
